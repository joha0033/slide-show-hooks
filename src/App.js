import React, { useState, useEffect, useReducer, useRef} from 'react';
import './App.css';
import {DATA as slides} from './data'

import {
  FaPlay,
  FaPause,
  FaChevronCircleRight,
  FaChevronCircleLeft,
  FaCircle
} from 'react-icons/fa'
export const SLIDE_DURATION = 3000
export const TIME_DURATION = 3000

export function Slide({
  isCurrent,
  takeFocus,
  slide,
  id, 
  title,
  children
}){
  let ref = useRef()
  useEffect(() => {
    if(isCurrent && takeFocus){
      ref.current.focus()
    }
  }, [isCurrent, takeFocus])

  return (
    <li
      ref={ref}
      aria-hidden={!isCurrent}
      tabIndex="-1"
      className="Slide"
      style={{backgroundImage:`url(${slide.hdurl})`}}
    >
      <div className="SlideContent">
        <h2 id={id} className="Title">
          {slide.title}
        </h2>
        <div className="Explanation">
          {children}
        </div>
      </div>
    </li>
   
  )
}
export function Slides(props){
  return (
    <ul className='Slides' {...props} />
  )
}

export function Carousel(props){
  return (
    <section className='Carousel' {...props} />
  )
}

export function Controls(props) {
  return <div className='Controls' {...props}/>
}

export function SlideNav(props) {
  return <div className='SlideNav' {...props}/>
}

export function SlideNavItem(props) {
  const { isCurrent } = props
  return isCurrent
    ? <button className='SlideNavItemOn' {...props}/>
    : <button className='SlideNavItemOff' {...props}/>
}

export function IconButton(props) {
  return <button className='IconButton' {...props}/>
}

export function SpacerGif({width}){
  return (
    <div
      style={{display:'inline-block', width}}
    />
  )
}

export function ProgressBar({animate, time}) {
  let progress = useProgress(animate, time)

  return (
  <div className='ProgressBar'>
    <div 
      style={{
        width: `${progress * 100}%`}}
      />
  </div>)
}

export const useProgress = (animate, time) => {
  let [ progress, setProgress ] = useState(0)

  useEffect(( ) => {
      if(animate){
          let rafId = null
          let start = null
          let step = timestamp => {
              if(!start) start = timestamp
              let progress = timestamp - start
              setProgress(progress)
              if(progress < time) {
                  rafId = requestAnimationFrame(step)
              }
          }
          rafId = requestAnimationFrame(step)
          return () => cancelAnimationFrame(rafId)
      }
  }, [time, animate])
  
  return animate 
      ? Math.min(progress/time, time)
      : 0
}

export function App() {
  let [state, dispatch] = useReducer((state, action) => {
    switch (action.type) {
      case 'PROGRESS':
      case 'NEXT': return {
        ...state,
        isPlaying: action.type === 'PROGRESS',
        takeFocus: false,
        currentIndex: (state.currentIndex + 1) % slides.length
      }
      case 'PREV': return {
        ...state,
        isPlaying: false,
        takeFocus: false,
        currentIndex: (state.currentIndex - 1 + slides.length) % slides.length
      }
      case 'PLAY': return {
        ...state,
        takeFocus: false,
        isPlaying: true,
      }
      case 'PAUSE': return {
        ...state,
        takeFocus: false,
        isPlaying: false,
      }
      case 'GOTO': return {
        ...state,
        takeFocus: true,
        currentIndex: action.index
      }
      default: return state
    }
  }, {
    currentIndex: 0,
    isPlaying: false,
    takeFocus: false
  })

  useEffect(() => {
    if(state.isPlaying){
      let timeout = setTimeout(() => {
        dispatch({type: 'PROGRESS'})
      }, SLIDE_DURATION)
      return () => {clearTimeout(timeout)}
    }
  }, [state.currentIndex, state.isPlaying])
    return (
      <div className="App">
        <header className="App-header">
          <Carousel>
            <Slides>
              {
                slides.map((slide, index) => {
                  return (
                    <Slide 
                      isCurrent={index===state.currentIndex}
                      key={index}
                      slide={slide}
                      takeFocus={state.takeFocus}
                      children={slide.explanation}
                    />
                  )
                })
              }
            
            </Slides>
            <SlideNav>
              {
                slides.map((slide, index) => {
                  return index === state.currentIndex
                    ? (<SlideNavItem
                      key={index}
                      children={<FaCircle style={{opacity: 1}}/>}
                      aria-label={`Slide ${index + 1}`}
                      onClick={() => {
                        dispatch({type: 'GOTO', index})
                      }}
                    />)
                    : (<SlideNavItem
                      key={index}
                      children={<FaCircle style={{opacity: .6}} />}
                      aria-label={`Slide ${index + 1}`}
                      onClick={() => {
                        dispatch({type: 'GOTO', index})
                      }}
                    />)
                })
              }
            </SlideNav>

            <Controls>
              {
                state.isPlaying
                  ? (
                    <IconButton
                      aria-label="pause"
                      onClick={() => {dispatch({type: 'PAUSE'})}}
                      children={<FaPause />}
                    />
                  )
                  : (
                    <IconButton
                      aria-label="Play"
                      onClick={() => {
                        dispatch({type: 'PLAY'})
                      }}
                      children={<FaPlay />}
                    />
                  )
              }
              <SpacerGif width="10px"/>
              <IconButton
                aria-label="Previous Slide"
                onClick={() => {
                  dispatch({type: 'PREV'})
                }}
                children={<FaChevronCircleLeft />}
              />
              <SpacerGif width="10px"/>
              <IconButton
                aria-label="Next Slide"
                onClick={() => {
                  dispatch({type: 'NEXT'})
                }}
                children={<FaChevronCircleRight />}
              />
            </Controls>

            <ProgressBar 
              key={state.currentIndex + state.isPlaying}
              time={TIME_DURATION}
              animate={state.isPlaying}
            />

          </Carousel>
        </header>
      </div>
    );
  }


export default App;
