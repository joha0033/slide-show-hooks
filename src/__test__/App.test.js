import React, { useRef, useEffect } from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer'
import {
  render
} from 'react-testing-library'

import {App, Slide, Slides} from '../App';
import { DATA } from '../data'

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

it('DATA is an object(an array) with length of 3', () => {
  expect(DATA.length).toEqual(3)
  expect(typeof DATA).toEqual('object')
});

describe('Slides component', () => {
  it('returns an <ul> tag ', () => {
    const tree = renderer
        .create(<Slides />)
        .toJSON()
    expect(tree.type).toEqual('ul')
    expect(tree.props.className).toEqual('Slides')
    expect(tree).toMatchSnapshot()
  })    
});



// describe('useProgress custom hook', () => {
//   it('returns an <ul> tag ', () => {
//     const tree = renderer
//         .create(<Slides />)
//         .toJSON()
//     expect(tree.type).toEqual('ul')
//     expect(tree.props.className).toEqual('Slides')
//     expect(tree).toMatchSnapshot()
//   })    
// });





// function SlidesTest() {
//   const div = document.createElement('div');
  // let state = {
  //   currentIndex: 0,
  //   takeFocus: true,
  // }
  // let index = 0
  // let slide = DATA[0]
// return (
//   <Slides>
//   {
//     slides.map((slide, index) => {
//       return (
//         <Slide 
//           isCurrent={index===state.currentIndex}
//           key={index}
//           slide={slide}
//           takeFocus={state.takeFocus}
//           children={slide.explanation}
//         />
//       )
//     })
//   }

// </Slides>
// )
// }