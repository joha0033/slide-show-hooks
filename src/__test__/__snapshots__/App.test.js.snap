// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slides component returns an <ul> tag  1`] = `
<ul
  className="Slides"
/>
`;
